accepted/future-releases/records/records-feature-specification.md

```diff
@@ -337,13 +347,84 @@ var c = cond ? a : b; // c has type `Record`.
 
 The greatest lower bound of records with different shapes is `Never`.
 
-### Type inference and promotion
+### Type inference
+
+Every record literal has a static type, which is associated with it via
+inference (there is no syntax for explicitly associating a specific static type
+with a record literal).  As with other constructs, we define type inference for
+record expressions with respect to a context type schema which is determined by
+the surrounding context of the inferred expression.  Unlike nominal classes (but
+like function literals) we choose to infer the most specific possible type for
+record literals, even if that type is more precise than the type specified by
+the context type schema.  This choice (as with function literals) reflects the
+fact that record types (like function types in Dart) are soundly variant: that
+is, the subtyping is properly covariant and requires no runtime checking.
+
+For convenience, we generally write function types with all named parameters in
+an unspecified canonical order, and similarly for the named fields of record
+types.  In all cases unless otherwise specifically called out, order of named
+parameters and fields is semantically irrelevant: any two types with the same
+named parameters (named fields, respectively) are considered the same type.
+
+Similarly, function and method invocations with named arguments and records with
+named field entries are written with their named entries in an unspecified
+canonical order and position.  Unless otherwise called out, position of named
+entries is semantically irrelevant, and all invocations and record literals with
+the same named entries (possibly in different orders or locations) and the same
+positional entries are considered equivalent.
+
+Given a type schema `K` and a record expression `E` of the general form `(e0,
+..., en, d0 : e{n+1}, ..., dm : e{m+n+1})` inference proceeds as follows.
+
+If `K` is a record type schema of the form `(K0, ..., Kn, {d0 : K{n+1}, ....,
+dm : K{m+n+1})` then:
+  - Each `ei` is inferred with context type schema `Ki` to have type `Ti`
+    - Let `Ri` be the greatest closure of `Ki`
+    - If `Ti` is a subtype of `Ri` then let `Ti'` be `Ti`
+    - Otherwise, if `Ti` is `dynamic`, then we insert an implicit cast on `ei`
+      to `Ri`, and let `Ti'` be `Ri`
+    - Otherwise, if `Ti` is coercible to `Ri` (via int-to-double conversion,
+      call method tearoff, or implicit generic instantiation), then we insert
+      the appropriate implicit coercion on `ei` to coerce it to `Ri`, and let
+      `Ti'` be `Ri`
```

Comment:
> (We can't use "int-to-double conversion" here. That must have happened to ei during its own inference. It's too late if that inference already inferred int.)

I agree that there is some oddity here, in that we haven't specified where coercions are done elsewhere.  That said, I don't understand your point here.  Can you say more about why we can't do this coercion here?  I think it is likely not *necessary* (because in practice it will have already been done), but it's certainly possible.  Two illustrative examples:

```
// Here, coercion has to happen inside the cascade.  `3..runtimeType` has type double, 
// no coercion implied by record inference 
(double d,) = (3..runtimeType,); 
// Here, we could say that the coercion has happened as part of inference of `3` (in which case no coercion
// is implied via record inference), or we could say that the coercion happens as part of inference of the record.  
(double d,) = (3,);
```

So I think that for int to double coercions, including them here is redundant/unnecessary, but I don't think it's wrong,  and in the absence of an actual specification of where int to double conversions are done, it felt safer to include them here.  For the remaining coercions, I think we do need to specify that they are done here, since I believe that they are not currently done eagerly.

> We need to say "via one or more of" for the other conversions.

I do not believe that we perform multiple conversions in one step (and the beliefs of the CFE and analyzer seem to be in line with mine).  Am I missing something here?

> I also think we should not let Ti be Ri here. It should instead be the actual, precise type of the coerced expression.

Makes sense.  Done.

Reply:
> in the absence of an actual specification of where int to double conversions are done

Where that coercion happens is completely specified. The semantics of a decimal literal with a context type which `double` is assignable to, and `int` is not, is a `double` value with static type `double`. 

The "double literal coercion" is not a value coercion. We never create an integer value at runtime, then convert it to double. Instead we change the semantics of the decimal number literal to directly evaluate to a double value. That's why it cannot happen later than at the evaluation of literal expression itself.

The distinction is important on native, where `double x = 0x10000000000000000;` is valid with value pow(2.0, 64), but going through an integer would be an error &mdash; `int x = 0x10000000000000000;` is a compile-time error because the numerical value of the literal doesn't fit in an `int`.

And in the same way, `double d = 0x3333333333333333;` is also an error because the numerical values doesn't fit in a double (so it's not interpreted as a double, and you get an error that `int` is not assignable to `double`).

(On the web, all values are double anyway.)

It *would* be invalid to do a double conversion at this point, because there is no conversion which is valid at the point where you already have an `int`-type subexpression.

Reply:
> I do not believe that we perform multiple conversions in one step

This works today, no warnings in DartPad or analyzer, and runs in the VM.
```dart
class Id {
  T call<T>(T value) => value;
}
int Function(int) f = Id();
void main() {
  print(f.runtimeType); // (int) => int
}
```
We perform both `.this` tear-off and `<int>` instantiation on the same expression.

In practice, we should reuse whatever specification we have for other expressions inside record literals too, so we get the same behavior. Or say we want the same behavior.
(We'll get rid of `.call` tear-off anyway, so we'll probably be down to one coercion when records launch anyway.)

Reply:
> This works today, no warnings in DartPad or analyzer, and runs in the VM.

Interesting.  This changed - I was testing in the 2.9 language version, and it was rejected.  But you are correct, this works on bleeding edge.  I don't know how or why this changed, did we actually make this change deliberately?  I don't think this is a good thing in general.  It's probably ok now given the limited set of coercions that we have, but if we ever add more coercions, I think we would almost certainly want to avoid multi-step coercions - it opens up all sorts of failure modes (e.g. ordering issues).

Reply:
Ack.  And just in the interest of clarification, from my perspective, "Otherwise, if Ti is coercible to Ri (via int-to-double conversion" in no way circumscribes the conversion to only look at the type.  A thing is only coercible via int-to-double conversion if the expression to be coerced is a literal.  It would be better if I had said `ei` is coercible from `Ti` to `Ri` I guess?

Reply:
Pedantically: No.
Mentioning `Ti` is the problem. The moment you know the static type of `ei`, type inference for `ei` has happened, and therefore it must *already* have been coerced to a double literal. If it hasn't, it's too late now anyway.

The other conversions (`.call`, generic instantiation, implicit downcast) are based on the static type and context type of an expression, and happens after inferring the static type (and everything type inference otherwise entails) on the expression. At runtime, the expression itself evaluates normally, then we do something to its value afterwards (`.call`/`<...>`/`as Ri`).

The decimal-numeral-to-double interpretation is based on the actual syntax of the expression and the context type, and happens *before* the static type is inferred. Then the static type becomes `double`. At runtime, we evaluate the literal directly to a double value. (And/or at compile time, since it's effectively constant.)

That's why there is no static type which is coercible to anything via int-to-double-conversion.

You can perhaps say, "if `ei` is coercible to `Ri` via ....", but it's easier to just not mention the (imprecisely named) int-to-double-conversion along with the other coercions that we may apply to the value of `e1`. Just make sure to propagate context types down, then the double interpretation will automatically be performed at the literal, by the literal. It's not something to worry about during upwards propagation at all. 

Basically, it's not a coercion at all. It's just that the static and dynamic semantics of an integer numeral depends on the expression's context type. Then it's just the plain semantics of the expression.

Reply:
> The moment you know the static type of ei, type inference for ei has happened, and therefore it must already have been coerced to a double literal. If it hasn't, it's too late now anyway.

Can you give a concrete example of what it means to be "too late" here?

Reply:
I guess that I'm saying that if an expression with context type `double` has been inferred a static type of `int`, then that is it. There is no later coercion which allows that `int` expression to be valid in a `double` context.

I don't think there is any example where the static type is `int`, the context type is `double`, and the expression is an integer literal, so we'd have to look at another expression than a literal.
```dart
int a = 1;
double x = a;
````
The static type of `a` is `int`, even in a `double` context. No coercion applies to that static type or that integer value.
It's simply a compile-time error to have an `int` expression in a `double` context.

(A remarkable number of expressions actually *do* work, including `double x = 1 + 1;`, which puts the second `1` in a `double` context.)

Or, to be short: "int-to-double" is not a coercion which applies to the value of an expression. It's simply a name we gave to the feature which changed the semantics of some integer literal expressions when they occur in a `double`-accepting context. Just like the set-or-map literal `{}` evaluates to either a set or a map depending on the context type.

The actual coercions are something we do to the resulting value of an expression, which is otherwise evaluated normally, in order to make the value match the expected context type. We insert behavior, statically and at runtime, to *make* the types match.

With integer literals, we could rename them to `integer-or-double-literal`, like  `map-or-set-literal`, and then we probably wouldn't be thinking about it as a coercion at all.
(Or we can think of it as an inferred change to the program, from integer literal to double literal, like the added inferred type arguments to `{}` which depends on the context type, but are effectively always there after type inference.)